File: Day #623/Solution.java
// Full text justification: greedily pack words per line, distribute spaces evenly
// with extra spaces favoring left gaps; last/single word left-justified. Time O(total chars).
import java.util.*;

public class Solution {
    static List<String> justify(String[] words, int k) {
        List<String> res = new ArrayList<>();
        int n = words.length, i = 0;
        while (i < n) {
            int j = i, lineLen = words[i].length();
            while (j + 1 < n && lineLen + 1 + words[j + 1].length() <= k) {
                j++;
                lineLen += 1 + words[j].length();
            }
            int gaps = j - i;
            StringBuilder line = new StringBuilder();
            if (gaps == 0) {
                line.append(words[i]);
                while (line.length() < k) line.append(' ');
            } else {
                int totalSpaces = k;
                for (int w = i; w <= j; w++) totalSpaces -= words[w].length();
                int base = totalSpaces / gaps, extra = totalSpaces % gaps;
                for (int w = i; w <= j; w++) {
                    line.append(words[w]);
                    if (w < j) {
                        int sp = base + (w - i < extra ? 1 : 0);
                        for (int s = 0; s < sp; s++) line.append(' ');
                    }
                }
            }
            res.add(line.toString());
            i = j + 1;
        }
        return res;
    }

    public static void main(String[] args) {
        String[] words = {"the","quick","brown","fox","jumps","over","the","lazy","dog"};
        for (String line : justify(words, 16))
            System.out.println('"' + line + '"');
    }
}

File: Day #623/solution.rs
// Full text justification: greedily pack words per line, distribute spaces evenly
// with extra spaces favoring left gaps; last/single word left-justified. Time O(total chars).

fn justify(words: &[&str], k: usize) -> Vec<String> {
    let mut res = Vec::new();
    let n = words.len();
    let mut i = 0;
    while i < n {
        let mut j = i;
        let mut line_len = words[i].len();
        while j + 1 < n && line_len + 1 + words[j + 1].len() <= k {
            j += 1;
            line_len += 1 + words[j].len();
        }
        let gaps = j - i;
        let mut line = String::new();
        if gaps == 0 {
            line.push_str(words[i]);
            line.push_str(&" ".repeat(k - words[i].len()));
        } else {
            let mut total_spaces = k;
            for w in i..=j {
                total_spaces -= words[w].len();
            }
            let base = total_spaces / gaps;
            let extra = total_spaces % gaps;
            for w in i..=j {
                line.push_str(words[w]);
                if w < j {
                    let sp = base + if w - i < extra { 1 } else { 0 };
                    line.push_str(&" ".repeat(sp));
                }
            }
        }
        res.push(line);
        i = j + 1;
    }
    res
}

fn main() {
    let words = ["the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog"];
    for line in justify(&words, 16) {
        println!("\"{}\"", line);
    }
}
